### PR TITLE
Wire batched einsum into training pipeline

### PR DESCRIPTION
**Pull request by zazabap**
## Summary
- `train_basis` now accepts `batch_size`, `optimizer` (:adam/:gradient_descent), and `device` (:cpu/:gpu) kwargs
- When batch_size > 1 with custom optimizer, pre-computes batched einsum codes once (TreeSA is expensive) and reuses for all epochs
- Add device abstraction layer (`to_device`/`to_cpu`) for GPU array management
- Add checkpointing support (save basis + loss history periodically)

## Files
- `src/training.jl` (modified, 354 insertions / 248 deletions)
- `test/training_tests.jl` (added batched training tests)
- `src/ParametricDFT.jl` (export `to_device`)

## Test plan
- [x] All existing training tests pass unchanged
- [x] Batched training with :adam optimizer and batch_size=3 works
- [x] Batched training with :gradient_descent works
- [x] batch_size=1 fallback still works

## PR chain
1. feature/riemannian-optimizers → main (PR #35)
2. feature/batched-einsum → PR #35 (PR #36)
3. **This PR** → PR #36
4. feature/cuda-extension → this PR

### CODE REVIEW

File: src/training.jl [hunk 1]
```diff
@@ -86,27 +149,65 @@ function _train_basis_core(
...
-                img_matrix, current_theta, M, optcode, inverse_code,
-                m, n, loss, steps_per_image
-            )
-
-            tensors = point2tensors(current_theta, M)
-            train_loss = loss_function(tensors, m, n, optcode, img_matrix, loss; inverse_code=inverse_code)
-            push!(epoch_losses, train_loss)
-            push!(step_train_losses, train_loss)
-            push!(loss_records, Dict{String, Any}("epoch" => epoch, "step" => idx, "loss" => train_loss))
+        # Train on batches
```
Comment (zazabap) [inline src/training.jl]: Should this be a helper function provide training on batch or not?
